### PR TITLE
Prevent emitting const const on C++ codegen when using a Slice with a reference

### PR DESCRIPTION
**Pull request by andfoy**
Fixes https://github.com/dtolnay/cxx/issues/830

### CODE REVIEW

File: gen/src/write.rs [hunk 1]
```diff
@@ -1205,7 +1205,14 @@ fn write_type(out: &mut OutFile, ty: &Type) {
         Type::SliceRef(slice) => {
             write!(out, "::rust::Slice<");
             if slice.mutability.is_none() {
-                write!(out, "const ");
+                match slice.inner {
+                    Type::Ref(_) => {
+                        write!(out, "");
```
Comment (andfoy) [inline gen/src/write.rs:1210]: @dtolnay Should I also add the const write here?

File: gen/src/write.rs [hunk 2]
```diff
@@ -1205,9 +1205,19 @@ fn write_type(out: &mut OutFile, ty: &Type) {
         Type::SliceRef(slice) => {
             write!(out, "::rust::Slice<");
             if slice.mutability.is_none() {
-                write!(out, "const ");
+                match slice.inner {
+                    Type::Ref(_) => {
```
Comment (andfoy) [inline gen/src/write.rs:1209]: @dtolnay Do I need to consider other types here?

Comment (dtolnay) [reply gen/src/write.rs:1209]: Yeah slices of pointers are still affected:

```console
$ echo '#[cxx::bridge] mod ffi { extern "Rust" { fn f(x: &[*const u8]); }}' | cargo run --manifest-path gen/cmd/Cargo.toml -- --header - | tail -1

void f(::rust::Slice<const const ::std::uint8_t *> x) noexcept;
```

**Review by dtolnay — CHANGES_REQUESTED** [review]
This turns out not to fix #830. :confused: 

```cpp
#include "rust/cxx.h"
void f(::rust::Slice<const ::std::uint8_t &const> x) noexcept;
```

```console
$ clang++-11 -std=c++20 main.cc -Wall
main.cc:2:44: error: 'const' qualifier may not be applied to a reference
void f(::rust::Slice<const ::std::uint8_t &const> x) noexcept;
                                           ^
$ g++-10 -std=c++20 main.cc -Wall
main.cc:2:44: error: ‘const’ qualifiers cannot be applied to ‘const uint8_t&’ {aka ‘const unsigned char&’}
  2 | void f(::rust::Slice<const ::std::uint8_t &const> x) noexcept;
    |                                            ^~~~~
```

**Comment by andfoy** [comment]
It seems that a container cannot hold references either: https://stackoverflow.com/questions/1155142/why-do-i-get-an-error-in-forming-reference-to-reference-type-map, instead one should wrap a reference around `std::reference_wrapper`, however, I don't know how to extend Slice to convert a reference into a `reference_wrapper`

**Comment by dtolnay** [comment]
Maybe with a [partial template specialization](https://en.cppreference.com/w/cpp/language/partial_specialization) in cxx.h which applies to instantiations of `rust::Slice<T &>`?

```cpp
template <typename T>
class Slice<T &> final {
    /*...*/
};
```

Alternatively it's possible we can fix this purely by using the existing partial specialized helpers that already exist in `<type_traits>`. For example instead of:

https://github.com/dtolnay/cxx/blob/064cf12b7850e4a3e2e09134eeea8fc619e60792/include/cxx.h#L170

we'd have:

```diff
-   T &operator[](std::size_t n) const noexcept;
+   std::add_lvalue_reference<T>::type operator[](std::size_t n) const noexcept;
```

According to https://en.cppreference.com/w/cpp/types/add_reference that has the semantics we want.

> If `T` is a function type that has no cv- or ref- qualifier or an object type, provides a member typedef `type` which is `T&`. If `T` is an rvalue reference to some type `U`, then `type` is `U&`. Otherwise, `type` is `T`.

In other words `rust::Slice<const Struct>::operator[]` would return `const Struct &`, whereas `rust::Slice<const Struct &>::operator[]` would *also* return `const Struct &`, because add_lvalue_reference is instantiated with a type that is already a reference.

**Comment by andfoy** [comment]
> In other words rust::Slice<const Struct>::operator[] would return const Struct &, whereas rust::Slice<const Struct &>::operator[] would also return const Struct &, because add_lvalue_reference is instantiated with a type that is already a reference.

Sounds like it would fix the issue, let me update the PR!

**Comment by andfoy** [comment]
I think it is necessary to update the generated cxx.cc file to consider `add_lvalue_reference` 

```
error: no declaration matches ‘T& rust::cxxbridge1::Slice<T>::operator[](std::size_t) const
```

**Comment by andfoy** [comment]
I tried this change locally, and it seems to work, although I have to use `std::reference_wrapper` on the signature

```rust
match &slice.inner {
    Type::Ref(r) => {
        // let inner_ref = r.inner;
        write!(out, "std::reference_wrapper<");
        write_type(out, &r.inner);
        write!(out, ">");
    }
    _ => {
        write!(out, "const ");
        write_type(out, &slice.inner);
    }
}
```

```c++
std::shared_ptr<CrossTensor> index(
    const std::shared_ptr<CrossTensor> &tensor,
    rust::Slice<std::reference_wrapper<CrossTensorRef>> indices);
```

**Comment by andfoy** [comment]
@dtolnay, do you prefer using `std::reference_wrapper` here?

File: gen/src/write.rs (hunk 2 above)
Comment (sara192) [reply gen/src/write.rs:1209]: > @dtolnay Do I need to consider other types here?

You

**Comment by TheZoq2** [comment]
Any chance of this being merged? I ran into the issue described in #830 so having a fix would be nice. I'd be happy to help out if some work is needed to get it done :)
